caas/kubernetes/provider/rbac.go

```diff
@@ -0,0 +1,462 @@
+	if k8serrors.IsAlreadyExists(err) {
+		return nil, errors.AlreadyExistsf("role %q", role.GetName())
+	}
+	return out, errors.Trace(err)
+}
+
+func (k *kubernetesClient) updateRole(role *rbacv1.Role) (*rbacv1.Role, error) {
+	out, err := k.client().RbacV1().Roles(k.namespace).Update(role)
+	if k8serrors.IsNotFound(err) {
+		return nil, errors.NotFoundf("role %q", role.GetName())
+	}
+	return out, errors.Trace(err)
+}
+
+func (k *kubernetesClient) ensureRole(role *rbacv1.Role) (*rbacv1.Role, error) {
+	out, err := k.updateRole(role)
+	if errors.IsNotFound(err) {
+		out, err = k.createRole(role)
+	}
+	return out, errors.Trace(err)
+}
+
+func (k *kubernetesClient) getRole(name string) (*rbacv1.Role, error) {
+	out, err := k.client().RbacV1().Roles(k.namespace).Get(name, v1.GetOptions{IncludeUninitialized: true})
+	if k8serrors.IsNotFound(err) {
+		return nil, errors.NotFoundf("role %q", name)
+	}
+	return out, errors.Trace(err)
+}
+
+func (k *kubernetesClient) deleteRole(name string) error {
+	err := k.client().RbacV1().Roles(k.namespace).Delete(name, &v1.DeleteOptions{
+		PropagationPolicy: &defaultPropagationPolicy,
+	})
+	if k8serrors.IsNotFound(err) {
+		return nil
+	}
+	return errors.Trace(err)
+}
+
+func (k *kubernetesClient) deleteRoles(appName string) error {
+	roleList, err := k.client().RbacV1().Roles(k.namespace).List(v1.ListOptions{
+		LabelSelector: applicationSelector(appName),
+	})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for _, s := range roleList.Items {
+		if err := k.deleteRole(s.GetName()); err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func (k *kubernetesClient) createClusterRole(cRole *rbacv1.ClusterRole) (*rbacv1.ClusterRole, error) {
+	out, err := k.client().RbacV1().ClusterRoles().Create(cRole)
+	if k8serrors.IsAlreadyExists(err) {
+		return nil, errors.AlreadyExistsf("cluster role %q", cRole.GetName())
+	}
+	return out, errors.Trace(err)
+}
+
+func (k *kubernetesClient) updateClusterRole(cRole *rbacv1.ClusterRole) (*rbacv1.ClusterRole, error) {
+	out, err := k.client().RbacV1().ClusterRoles().Update(cRole)
+	if k8serrors.IsNotFound(err) {
+		return nil, errors.NotFoundf("cluster role %q", cRole.GetName())
+	}
+	return out, errors.Trace(err)
+}
+
+func (k *kubernetesClient) ensureClusterRole(cRole *rbacv1.ClusterRole) (*rbacv1.ClusterRole, error) {
+	out, err := k.updateClusterRole(cRole)
+	if errors.IsNotFound(err) {
+		out, err = k.createClusterRole(cRole)
+	}
+	return out, errors.Trace(err)
+}
+
+func (k *kubernetesClient) getClusterRole(name string) (*rbacv1.ClusterRole, error) {
+	out, err := k.client().RbacV1().ClusterRoles().Get(name, v1.GetOptions{IncludeUninitialized: true})
+	if k8serrors.IsNotFound(err) {
+		return nil, errors.NotFoundf("cluster role %q", name)
+	}
+	return out, errors.Trace(err)
+}
+
+func (k *kubernetesClient) deleteClusterRole(name string) error {
+	err := k.client().RbacV1().ClusterRoles().Delete(name, &v1.DeleteOptions{
```

Comment:
Seeing as cluster roles are global, we should validate the cluster role is managed by this controller.

Reply:
as discussed, we already ensure this by labelling the resources.